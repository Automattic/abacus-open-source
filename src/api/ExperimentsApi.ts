import _ from 'lodash'
import * as yup from 'yup'

import {
  ExperimentFull,
  ExperimentFullNew,
  experimentFullNewOutboundSchema,
  experimentFullNewSchema,
  experimentFullSchema,
  ExperimentSummary,
  experimentSummaryResponse,
  MetricAssignmentNew,
  metricAssignmentNewOutboundSchema,
  metricAssignmentNewSchema,
  Status,
  yupPick,
} from 'src/lib/schemas'
import { isDebugMode } from 'src/utils/general'

import { fetchApi } from './utils'

/**
 * Attempts to create a new experiment.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function create(newExperiment: ExperimentFullNew): Promise<ExperimentFull> {
  const validatedNewExperiment = await experimentFullNewSchema.validate(newExperiment, { abortEarly: false })
  const outboundNewExperiment = experimentFullNewOutboundSchema.cast(validatedNewExperiment)
  return await experimentFullSchema.validate(await fetchApi('POST', '/experiments', outboundNewExperiment))
}

/**
 * Attempts to PUT an experiment, overwriting the existing experiment.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function put(experimentId: number, experiment: ExperimentFullNew): Promise<ExperimentFull> {
  const validatedExperiment = await experimentFullNewSchema.validate(experiment, { abortEarly: false })
  const outboundExperiment = experimentFullNewOutboundSchema.cast(validatedExperiment)
  return await experimentFullSchema.validate(await fetchApi('PUT', `/experiments/${experimentId}`, outboundExperiment))
}

/**
 * Attempts to patch an experiment.
 *
 * Doesn't work with nested fields.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function patch(experimentId: number, experimentPatch: Partial<ExperimentFull>): Promise<ExperimentFull> {
  // We dynamically construct a schema for validation, but we do this simply and shallowly
  const dynamicValidationSchema = yupPick(experimentFullSchema, Object.keys(experimentPatch))
  const validatedExperimentPatch = await dynamicValidationSchema.validate(experimentPatch, { abortEarly: false })

  // Similarly we dynamically construct a schema for outbound casting
  const dynamicOutboundCastSchema = yupPick(experimentFullSchema, Object.keys(experimentPatch)).snakeCase()
  const outboundExperimentPatch = dynamicOutboundCastSchema.cast(validatedExperimentPatch)

  return await experimentFullSchema.validate(
    await fetchApi('PATCH', `/experiments/${experimentId}`, outboundExperimentPatch),
  )
}

async function changeStatus(experimentId: number, status: Status): Promise<void> {
  await fetchApi('PUT', `/experiments/${experimentId}/status`, { status })
}

/**
 * Attempts to assign a metric to an experiment.
 *
 * Metric Assignment is a bit funky right now, so we do it here.
 * In particular it require passing in all existing metricAssignments.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @param experiment A full experiment object, needed to fill in the existing metricAssignments, will be removed in the future.
 * @param metricAssignment The new metricAssignment.
 */
async function assignMetric(
  experiment: ExperimentFull,
  metricAssignment: MetricAssignmentNew,
): Promise<ExperimentFull> {
  const validatedMetricAssignment = await metricAssignmentNewSchema.validate(metricAssignment)
  const outboundMetricAssignment = metricAssignmentNewOutboundSchema.cast(validatedMetricAssignment)

  // The backend replaces all the metricAssignments, so we format it as metricAssignmentNewOutbound.
  const outboundExistingMetricAssignments = yup
    .array(metricAssignmentNewOutboundSchema)
    .cast(experiment.metricAssignments.map((metricAssignment) => _.omit(metricAssignment, 'metricAssignmentId')))
  // istanbul ignore next; Shouldn't occur
  if (!outboundExistingMetricAssignments) {
    throw new Error('Something went wrong while transforming existing metricAssignments to outbound form.')
  }

  const metricAssignments = [...outboundExistingMetricAssignments, outboundMetricAssignment]

  return await experimentFullSchema.validate(
    await fetchApi('PATCH', `/experiments/${experiment.experimentId}`, {
      metric_assignments: metricAssignments,
    }),
  )
}

/**
 * Finds all the available experiments.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findAll(): Promise<ExperimentSummary[]> {
  // istanbul ignore next; debug only
  const { experiments } = await experimentSummaryResponse.validate(
    await fetchApi('GET', isDebugMode() ? '/experiments?debug=true' : '/experiments'),
    { abortEarly: false },
  )
  return experiments
}

/**
 * Fetches the experiment with full details.
 *
 * @param id - The ID of the experiment to fetch.
 */
async function findById(id: number): Promise<ExperimentFull> {
  return await experimentFullSchema.validate(await fetchApi('GET', `/experiments/${id}`), { abortEarly: false })
}

const ExperimentsApi = {
  create,
  put,
  patch,
  assignMetric,
  changeStatus,
  findAll,
  findById,
}

export default ExperimentsApi
