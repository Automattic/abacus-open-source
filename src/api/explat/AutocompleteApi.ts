import { StatusCodes } from 'http-status-codes'

import { fetchApi } from 'src/api/explat/utils'
import { AutocompleteItem, autocompleteSchema, eventDetailsSchema } from 'src/lib/explat/schemas'

import HttpResponseError from '../HttpResponseError'

async function getCompletion(name: string) {
  return await autocompleteSchema.validate(await fetchApi('GET', `/autocomplete/${name}`), { abortEarly: false })
}

export async function getUserCompletions(): Promise<AutocompleteItem[]> {
  return (await getCompletion('users')).completions
}

export async function getEventNameCompletions(): Promise<AutocompleteItem[]> {
  return (await getCompletion('events')).completions
}

export async function getPropNameCompletions(eventName: string): Promise<AutocompleteItem[] | null> {
  if (!eventName) {
    throw new Error('No eventName to getPropNameCompletions of.')
  }

  try {
    const apiResponse = await eventDetailsSchema.validate(await fetchApi('GET', `/autocomplete/events/${eventName}`))
    return apiResponse.props.map((p) => ({
      name: p.name,
      value: p.name,
    }))
  } catch (error) {
    // istanbul ignore else; Forced to use the else here to prevent two istanbul ignores
    if (error instanceof HttpResponseError && error.status === StatusCodes.NOT_FOUND) {
      return null
    } else {
      throw error
    }
  }
}
