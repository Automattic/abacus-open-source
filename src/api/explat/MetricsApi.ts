import _ from 'lodash'
import * as yup from 'yup'

import {
  Metric,
  MetricNew,
  MetricNewOutbound,
  metricNewOutboundSchema,
  metricNewSchema,
  metricSchema,
} from 'src/lib/explat/schemas'

import { fetchApi } from './utils'

/**
 * Transforms a metric into an outbound form.
 */

function outboundTransformMetric(metricNew: MetricNew): MetricNewOutbound {
  const metricDataForRequest = {
    ...metricNew,
    tags: metricNew.tags.map((tagId) => ({ tagId })),
  }
  return metricNewOutboundSchema.cast(metricDataForRequest)
}

/**
 * Attempts to create a new metric.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function create(newMetric: MetricNew): Promise<Metric> {
  const validatedNewMetric = await metricNewSchema.validate(newMetric, { abortEarly: false })
  const outboundNewMetric = outboundTransformMetric(validatedNewMetric)
  return await metricSchema.validate(await fetchApi('POST', '/metrics', outboundNewMetric))
}

/**
 * Attempts to put a new metric.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function put(metricId: number, newMetric: MetricNew): Promise<Metric> {
  // istanbul ignore next; Shouldn't happen
  if (!_.isNumber(metricId)) {
    throw new Error('Invalid metricId.')
  }
  const validatedNewMetric = await metricNewSchema.validate(newMetric, { abortEarly: false })
  const outboundNewMetric = outboundTransformMetric(validatedNewMetric)
  return await metricSchema.validate(await fetchApi('PUT', `/metrics/${metricId}`, outboundNewMetric))
}

/**
 * Finds all the available metrics.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findAll(options: { includeDebug: boolean } = { includeDebug: false }): Promise<Metric[]> {
  // istanbul ignore next; debug only
  const { metrics } = await yup
    .object({ metrics: yup.array(metricSchema).defined() })
    .defined()
    .validate(await fetchApi('GET', options.includeDebug ? '/metrics?debug=true' : '/metrics'), {
      abortEarly: false,
    })
  return metrics
}

/**
 * Find the metric by ID.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findById(metricId: number): Promise<Metric> {
  return await metricSchema.validate(await fetchApi('GET', `/metrics/${metricId}`), { abortEarly: false })
}

const MetricsApi = {
  create,
  put,
  findAll,
  findById,
}

export default MetricsApi
