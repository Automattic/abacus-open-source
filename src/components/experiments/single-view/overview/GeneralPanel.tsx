import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  InputAdornment,
  Link,
  Paper,
  Toolbar,
  Tooltip,
  Typography,
} from '@material-ui/core'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import { Edit } from '@material-ui/icons'
import * as dateFns from 'date-fns'
import { Field, Formik } from 'formik'
import { TextField } from 'formik-material-ui'
import _ from 'lodash'
import { useSnackbar } from 'notistack'
import React, { useState } from 'react'
import * as yup from 'yup'

import ExperimentsApi from 'src/api/ExperimentsApi'
import { serverErrorMessage } from 'src/api/HttpResponseError'
import ExperimentStatus from 'src/components/experiments/ExperimentStatus'
import DatetimeText from 'src/components/general/DatetimeText'
import LabelValueTable from 'src/components/general/LabelValueTable'
import { ExperimentFull, experimentFullSchema, Status, yupPick } from 'src/lib/schemas'
import { formatIsoDate } from 'src/utils/time'

import { AssignmentCacheStatusToHuman } from '../../../../lib/experiments'
import LoadingButtonContainer from '../../../general/LoadingButtonContainer'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    to: {
      marginLeft: theme.spacing(2),
      marginRight: theme.spacing(2),
      fontFamily: theme.custom.fonts.monospace,
      color: theme.palette.grey[600],
    },
    title: {
      flexGrow: 1,
    },
    row: {
      margin: theme.spacing(5, 0),
      display: 'flex',
      alignItems: 'center',
      '&:first-of-type': {
        marginTop: theme.spacing(3),
      },
    },
    datePicker: {
      '& input:invalid': {
        // Fix the native date-picker placeholder text colour
        color: theme.palette.text.hint,
      },
    },
    monospace: {
      fontFamily: theme.custom.fonts.monospace,
    },
  }),
)

/**
 * Renders the general information of an experiment in a panel component.
 *
 * @param experiment - The experiment with the general information.
 * @param experimentReloadRef - Ref to reload the experiment.
 */
function GeneralPanel({
  className,
  experiment,
  experimentReloadRef,
}: {
  className?: string
  experiment: ExperimentFull
  experimentReloadRef: React.MutableRefObject<() => void>
}): JSX.Element {
  const classes = useStyles()
  const data = [
    {
      label: 'Status',
      value: <ExperimentStatus status={experiment.status} />,
    },
    {
      label: 'Dates',
      value: (
        <>
          <DatetimeText datetime={experiment.startDatetime} excludeTime />
          <span className={classes.to}>to</span>
          <DatetimeText datetime={experiment.endDatetime} excludeTime />
        </>
      ),
    },
    { label: 'Description', value: <span className={classes.monospace}>{experiment.description}</span> },
    { label: 'Owner', value: <span className={classes.monospace}>{experiment.ownerLogin}</span> },
    {
      label: 'P2 Link',
      value: (
        <Link href={experiment.p2Url} rel='noopener noreferrer' target='_blank' className={classes.monospace}>
          {experiment.p2Url}
        </Link>
      ),
    },
    {
      label: 'Assignment Cache Entry',
      value: (
        <span className={classes.monospace}>
          {AssignmentCacheStatusToHuman[experiment.assignmentCacheStatus]}&nbsp;(
          <Link
            href="https://github.com/Automattic/experimentation-platform/wiki/Experimenter's-Guide#the-file-system-cache"
            rel='noopener noreferrer'
            target='_blank'
            underline='always'
          >
            learn more
          </Link>
          )
        </span>
      ),
    },
  ]

  // Edit Modal
  const { enqueueSnackbar } = useSnackbar()
  const [isEditing, setIsEditing] = useState<boolean>(false)
  const generalEditInitialExperiment = {
    ..._.pick(experiment, ['description', 'ownerLogin']),
    endDatetime: formatIsoDate(experiment.endDatetime),
    // Needed for endDatetime validation
    startDatetime: experiment.startDatetime,
  }
  const canEdit = experiment.status !== Status.Staging
  const canEditEndDate = experiment.status === Status.Running
  const generalEditValidationSchema = yupPick(experimentFullSchema, ['description', 'ownerLogin']).shape({
    ...(canEditEndDate && {
      // We need to ensure the end date is in the future
      endDatetime: (yup.reach(experimentFullSchema, 'endDatetime') as unknown as yup.MixedSchema).test(
        'future-end-date',
        'End date (UTC) must be in the future.',
        // We need to refer to new Date() instead of using dateFns.isFuture so MockDate works with this in the tests.
        (date) => dateFns.isBefore(new Date(), date),
      ),
    }),
  })
  const onEdit = () => setIsEditing(true)
  const onCancelEdit = () => setIsEditing(false)
  const onSubmitEdit = async (formData: { experiment: typeof generalEditInitialExperiment }) => {
    try {
      const experimentPatch = _.pick(
        formData.experiment,
        canEditEndDate ? ['description', 'ownerLogin', 'endDatetime'] : ['description', 'ownerLogin'],
      )
      await ExperimentsApi.patch(experiment.experimentId, experimentPatch as unknown as Partial<ExperimentFull>)
      enqueueSnackbar('Experiment Updated!', { variant: 'success' })
      experimentReloadRef.current()
      setIsEditing(false)
    } catch (e) /* istanbul ignore next; Shouldn't happen */ {
      console.error(e)
      enqueueSnackbar(`Oops! Something went wrong while trying to update your experiment. ${serverErrorMessage(e)}`, {
        variant: 'error',
      })
    }
  }

  return (
    <Paper className={className}>
      <Toolbar>
        <Typography className={classes.title} color='textPrimary' variant='h3'>
          General
        </Typography>
        <Tooltip title={canEdit ? '' : 'Use "Edit in Wizard" for staging experiments.'}>
          <div>
            <Button onClick={onEdit} variant='outlined' disabled={!canEdit} aria-label='Edit Experiment General Data'>
              <Edit />
              Edit
            </Button>
          </div>
        </Tooltip>
      </Toolbar>
      <LabelValueTable data={data} />
      <Dialog open={isEditing} fullWidth aria-labelledby='edit-experiment-general-form-dialog-title'>
        <DialogTitle id='edit-experiment-general-form-dialog-title'>Edit Experiment: General</DialogTitle>
        <Formik
          initialValues={{ experiment: generalEditInitialExperiment }}
          validationSchema={yup.object({ experiment: generalEditValidationSchema })}
          onSubmit={onSubmitEdit}
        >
          {(formikProps) => (
            <form onSubmit={formikProps.handleSubmit} noValidate>
              <DialogContent>
                <div className={classes.row}>
                  <Field
                    component={TextField}
                    name='experiment.description'
                    id='experiment.description'
                    label='Experiment description'
                    placeholder='Monthly vs. yearly pricing'
                    helperText='State your hypothesis.'
                    variant='outlined'
                    fullWidth
                    required
                    multiline
                    rows={4}
                    InputLabelProps={{
                      shrink: true,
                    }}
                  />
                </div>

                <div className={classes.row}>
                  <Field
                    component={TextField}
                    className={classes.datePicker}
                    name='experiment.endDatetime'
                    id='experiment.endDatetime'
                    label='End date'
                    disabled={!canEditEndDate}
                    helperText={
                      canEditEndDate ? 'Use the UTC timezone.' : `Cannot be changed as the experiment has finished.`
                    }
                    type='date'
                    variant='outlined'
                    fullWidth
                    required
                    InputLabelProps={{
                      shrink: true,
                    }}
                  />
                </div>

                <div className={classes.row}>
                  <Field
                    component={TextField}
                    name='experiment.ownerLogin'
                    id='experiment.ownerLogin'
                    label='Owner'
                    placeholder='wp_username'
                    helperText='Use WordPress.com username.'
                    variant='outlined'
                    fullWidth
                    required
                    InputProps={{
                      startAdornment: <InputAdornment position='start'>@</InputAdornment>,
                    }}
                    InputLabelProps={{
                      shrink: true,
                    }}
                  />
                </div>
              </DialogContent>
              <DialogActions>
                <Button onClick={onCancelEdit} color='primary'>
                  Cancel
                </Button>
                <LoadingButtonContainer isLoading={formikProps.isSubmitting}>
                  <Button
                    type='submit'
                    variant='contained'
                    color='secondary'
                    disabled={formikProps.isSubmitting || !formikProps.isValid}
                  >
                    Save
                  </Button>
                </LoadingButtonContainer>
              </DialogActions>
            </form>
          )}
        </Formik>
      </Dialog>
    </Paper>
  )
}

export default GeneralPanel
