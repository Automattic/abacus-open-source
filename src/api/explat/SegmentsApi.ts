import * as yup from 'yup'

import { Segment, segmentSchema } from 'src/lib/explat/schemas'

import { fetchApi } from './utils'

/**
 * Finds all the available segments.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findAll(): Promise<Segment[]> {
  const { segments } = await yup
    .object({ segments: yup.array(segmentSchema).defined() })
    .defined()
    .validate(await fetchApi('GET', '/segments'), { abortEarly: false })
  return segments
}

const SegmentsApi = {
  findAll,
}

export default SegmentsApi
