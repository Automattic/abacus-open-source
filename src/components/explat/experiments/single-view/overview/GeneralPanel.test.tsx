/* eslint-disable @typescript-eslint/require-await, no-irregular-whitespace */
import { fireEvent, screen, waitFor, waitForElementToBeRemoved } from '@testing-library/react'
import { noop } from 'lodash'
import MockDate from 'mockdate'
import React from 'react'

import ExperimentsApi from 'src/api/explat/ExperimentsApi'
import { Status } from 'src/lib/explat/schemas'
import Fixtures from 'src/test-helpers/fixtures'
import { changeFieldByRole, render } from 'src/test-helpers/test-utils'

import GeneralPanel from './GeneralPanel'

MockDate.set('2020-07-21')

jest.mock('src/api/explat/ExperimentsApi')
const mockedExperimentsApi = ExperimentsApi as jest.Mocked<typeof ExperimentsApi>

const experimentReloadRef: React.MutableRefObject<() => void> = { current: noop }

test('renders as expected', () => {
  const experiment = Fixtures.createExperimentFull()
  const { container } = render(<GeneralPanel experiment={experiment} experimentReloadRef={experimentReloadRef} />)

  expect(container).toMatchInlineSnapshot(`
    <div>
      <div
        class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        >
          <h3
            class="MuiTypography-root makeStyles-title-2 MuiTypography-h3 MuiTypography-colorTextPrimary"
          >
            General
          </h3>
          <div
            class=""
            title="Use \\"Edit in Wizard\\" for staging experiments."
          >
            <button
              aria-label="Edit Experiment General Data"
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined Mui-disabled Mui-disabled"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
                Edit
              </span>
            </button>
          </div>
        </div>
        <table
          class="MuiTable-root"
        >
          <tbody
            class="MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="cell"
                scope="row"
              >
                Status
              </th>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <span
                  class="makeStyles-root-6"
                >
                  <span
                    class="makeStyles-statusDot-11 makeStyles-staging-9"
                  />
                   
                  staging
                </span>
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="cell"
                scope="row"
              >
                Dates
              </th>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <span
                  class="makeStyles-root-12"
                  title="20/09/2020, 20:00:00"
                >
                  2020-09-21
                </span>
                <span
                  class="makeStyles-to-1"
                >
                  to
                </span>
                <span
                  class="makeStyles-root-12"
                  title="20/11/2020, 19:00:00"
                >
                  2020-11-21
                </span>
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="cell"
                scope="row"
              >
                Description
              </th>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <span
                  class="makeStyles-monospace-5"
                >
                  Experiment with things. Change stuff. Profit.
                </span>
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="cell"
                scope="row"
              >
                Owner
              </th>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <span
                  class="makeStyles-monospace-5"
                >
                  owner-nickname
                </span>
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="cell"
                scope="row"
              >
                P2 Link
              </th>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-monospace-5 MuiTypography-colorPrimary"
                  href="https://wordpress.com/experiment_1"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  https://wordpress.com/experiment_1
                </a>
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="cell"
                scope="row"
              >
                Assignment Cache Entry
              </th>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                <span
                  class="makeStyles-monospace-5"
                >
                  ✅ Fresh: Production cache is up to date, but manual sandbox updates may be needed
                   (
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineAlways MuiTypography-colorPrimary"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    learn more
                  </a>
                  )
                </span>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  `)
})

test('renders as expected for experiment with null datetimes', () => {
  const experiment = Fixtures.createExperimentFull()
  const { container } = render(
    <GeneralPanel
      experiment={{ ...experiment, startDatetime: null, endDatetime: null }}
      experimentReloadRef={experimentReloadRef}
    />,
  )

  expect(container).toMatchSnapshot()
})

test('opens, submits and cancels edit dialog with running experiment', async () => {
  const experiment = Fixtures.createExperimentFull({ status: Status.Running })
  render(<GeneralPanel experiment={experiment} experimentReloadRef={experimentReloadRef} />)

  mockedExperimentsApi.patch.mockReset()
  mockedExperimentsApi.patch.mockImplementationOnce(async () => experiment)

  const editButton = screen.getByRole('button', { name: /Edit/ })
  fireEvent.click(editButton)

  await waitFor(() => screen.getByRole('button', { name: /Save/ }))

  await changeFieldByRole('textbox', /Experiment description/, 'Edited description.')
  // This date was picked as it is after the fixture start date.
  fireEvent.change(screen.getByLabelText(/End date/), { target: { value: '2020-10-20' } })
  await changeFieldByRole('textbox', /Owner/, 'changed_owner-nickname')

  const saveButton = screen.getByRole('button', { name: /Save/ })
  fireEvent.click(saveButton)
  await waitForElementToBeRemoved(saveButton)

  expect(mockedExperimentsApi.patch).toHaveBeenCalledTimes(1)
  expect(mockedExperimentsApi.patch).toMatchInlineSnapshot(`
    [MockFunction] {
      "calls": Array [
        Array [
          1,
          Object {
            "description": "Edited description.",
            "endDatetime": "2020-10-20",
            "ownerLogin": "changed_owner-nickname",
            "p2Url": "https://wordpress.com/experiment_1",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": Promise {},
        },
      ],
    }
  `)

  fireEvent.click(editButton)

  await waitFor(() => screen.getByRole('button', { name: /Cancel/ }))

  const cancelButton = screen.getByRole('button', { name: /Cancel/ })
  fireEvent.click(cancelButton)
  await waitForElementToBeRemoved(cancelButton)

  expect(mockedExperimentsApi.patch).toHaveBeenCalledTimes(1)
})

test('checks edit dialog does not allow end datetime changes with disabled experiment', async () => {
  const experiment = Fixtures.createExperimentFull({ status: Status.Disabled })
  render(<GeneralPanel experiment={experiment} experimentReloadRef={experimentReloadRef} />)

  mockedExperimentsApi.patch.mockReset()
  mockedExperimentsApi.patch.mockImplementationOnce(async () => experiment)

  const editButton = screen.getByRole('button', { name: /Edit/ })
  fireEvent.click(editButton)

  await waitFor(() => screen.getByRole('button', { name: /Save/ }))

  await changeFieldByRole('textbox', /Experiment description/, 'Edited description.')
  expect(screen.getByLabelText(/End date/)).toBeDisabled()
  await changeFieldByRole('textbox', /Owner/, 'changed_owner-nickname')
  await changeFieldByRole('textbox', /P2 post URL/, 'https://wordpress.com/experiment_changed')

  const saveButton = screen.getByRole('button', { name: /Save/ })
  fireEvent.click(saveButton)
  await waitForElementToBeRemoved(saveButton)

  expect(mockedExperimentsApi.patch).toHaveBeenCalledTimes(1)
  expect(mockedExperimentsApi.patch).toMatchInlineSnapshot(`
    [MockFunction] {
      "calls": Array [
        Array [
          1,
          Object {
            "description": "Edited description.",
            "ownerLogin": "changed_owner-nickname",
            "p2Url": "https://wordpress.com/experiment_changed",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": Promise {},
        },
      ],
    }
  `)
})
