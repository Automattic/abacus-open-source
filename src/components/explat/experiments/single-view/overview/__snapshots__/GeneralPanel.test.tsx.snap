// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected for experiment with null datetimes 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <h3
        class="MuiTypography-root makeStyles-title-14 MuiTypography-h3 MuiTypography-colorTextPrimary"
      >
        General
      </h3>
      <div
        class=""
        title="Use \\"Edit in Wizard\\" for staging experiments."
      >
        <button
          aria-label="Edit Experiment General Data"
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
            </svg>
            Edit
          </span>
        </button>
      </div>
    </div>
    <table
      class="MuiTable-root"
    >
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            Status
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            <span
              class="makeStyles-root-18"
            >
              <span
                class="makeStyles-statusDot-23 makeStyles-staging-21"
              />
               
              staging
            </span>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            Dates
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            -
            <span
              class="makeStyles-to-13"
            >
              to
            </span>
            -
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            Description
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            <span
              class="makeStyles-monospace-17"
            >
              Experiment with things. Change stuff. Profit.
            </span>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            Owner
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            <span
              class="makeStyles-monospace-17"
            >
              owner-nickname
            </span>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            P2 Link
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-monospace-17 MuiTypography-colorPrimary"
              href="https://wordpress.com/experiment_1"
              rel="noopener noreferrer"
              target="_blank"
            >
              https://wordpress.com/experiment_1
            </a>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            Assignment Cache Entry
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            <span
              class="makeStyles-monospace-17"
            >
              ✅ Fresh: Production cache is up to date, but manual sandbox updates may be needed
               (
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineAlways MuiTypography-colorPrimary"
                rel="noopener noreferrer"
                target="_blank"
              >
                learn more
              </a>
              )
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
